File: Seccion_III_Regression/Multiple_Lineal_Regression/multiple_lineal_regression.py
#Regresión lineal múltiple

#Importamos las librerias
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

#Importamos el dataset
dataset = pd.read_csv('50_Startups.csv')
X = dataset.iloc[:,:-1].values
y = dataset.iloc[:,4].values

#Tranformación de datos categóricos a variables dummies 
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.compose import make_column_transformer

labelencoder_X = LabelEncoder()
X[:, 3] = labelencoder_X.fit_transform(X[:, 3])
onehotencoder = make_column_transformer((OneHotEncoder(), [3]), remainder = "passthrough")
X = onehotencoder.fit_transform(X)

#Para evitar la trampa de las variables dummies
X = X[:, 1:]

#División del dataset en conj. de datos para la fase de entrenamiento yla de testing 
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

# Ajuste del modelo de Regresión lineal múltiple con el conjunto de entrenamiento
from sklearn.linear_model import LinearRegression
regression = LinearRegression()
regression.fit(X_train, y_train)

# Predicción de los resultados en el conjunto de testing
y_pred = regression.predict(X_test)

# Construcción del modelo óptimo de RLM utilizando la técnica de la Eliminación hacia atrás
import statsmodels.api as sm
X = np.append(arr = np.ones((50,1)).astype(int), values = X, axis = 1)
SL = 0.05

#Se ha añadido el modificador .tolist() al X_opt para adaptarse a Python 3.7

X_opt = X[:, [0, 1, 2, 3, 4, 5]]
regression_OLS = sm.OLS(endog = y, exog = X_opt.tolist()).fit()
regression_OLS.summary()

X_opt = X[:, [0, 1, 3, 4, 5]]
regression_OLS = sm.OLS(endog = y, exog = X_opt.tolist()).fit()
regression_OLS.summary()

X_opt = X[:, [0, 3, 5]]
regression_OLS = sm.OLS(endog = y, exog = X_opt.tolist()).fit()
regression_OLS.summary()

File: Seccion_III_Regression/Multiple_Lineal_Regression/multiple_lineal_regression.R
#Regresion lineal multiple

#Importamos el dataset
dataset = read.csv("50_Startups.csv")

#Transformación (codificacion) de datos categoricos a numericos

dataset$State = factor(dataset$State, 
                         levels = c("New York", "California", "Florida"),
                         labels = c(1, 2, 3)
)



#Division de los datos para la fase de entrenamiento y para la fase de testing
install.packages("caTools")
library(caTools)

set.seed(123)
split = sample.split(dataset$Profit, SplitRatio = 0.8)
training_set = subset(dataset, split == TRUE)
testing_set = subset(dataset, split == FALSE)

#Ajuste del modelo de regresión lineal múltiple con el conjunto de entrenamiento

regression = lm(formula = Profit ~ .,
                data = training_set)

#Predecimos los resultados con el conjunto de testing
y_pred = predict(regression, newdata = testing_set)

# Construcción de un modelo óptimo con el método de la Eliminación hacia atrás
SL = 0.05
regression = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State,
                data = dataset)
summary(regression)


regression = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend,
                data = dataset)
summary(regression)


regression = lm(formula = Profit ~ R.D.Spend + Marketing.Spend,
                data = dataset)
summary(regression)


regression = lm(formula = Profit ~ R.D.Spend,
                data = dataset)
summary(regression)










